commercetools/provider.go

```diff
@@ -17,52 +18,57 @@ func Provider() terraform.ResourceProvider {
+				Description: "The OAuth Client ID for a commercetools platform project. https://docs.commercetools.com/http-api-authorization",
+				Sensitive:   true,
 			},
 			"client_secret": {
-				Type:     schema.TypeString,
-				Optional: true,
-				DefaultFunc: schema.MultiEnvDefaultFunc([]string{
-					"CTP_CLIENT_SECRET",
-				}, nil),
-				Description: "CommercesTools Client Secret",
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("CTP_CLIENT_SECRET", nil),
+				Description: "The OAuth Client Secret for a commercetools platform project. https://docs.commercetools.com/http-api-authorization",
+				Sensitive:   true,
 			},
 			"project_key": {
-				Type:     schema.TypeString,
-				Optional: true,
-				DefaultFunc: schema.MultiEnvDefaultFunc([]string{
-					"CTP_PROJECT_KEY",
-				}, nil),
-				Description: "CommercesTools Project key",
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("CTP_PROJECT_KEY", nil),
+				Description: "The project key of commercetools platform project. https://docs.commercetools.com/getting-started",
+				Sensitive:   true,
 			},
-			"token_url": {
-				Type:     schema.TypeString,
-				Optional: true,
-				DefaultFunc: schema.MultiEnvDefaultFunc([]string{
-					"CTP_AUTH_URL",
-				}, "https://auth.sphere.io"),
-				Description: "CommercesTools Token URL",
+			"scopes": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("CTP_SCOPES", nil),
+				Description: "A list as string of OAuth scopes assigned to a project key, to access resources in a commercetools platform project. https://docs.commercetools.com/http-api-authorization",
 			},
 			"api_url": {
-				Type:     schema.TypeString,
-				Optional: true,
-				DefaultFunc: schema.MultiEnvDefaultFunc([]string{
-					"CTP_API_URL",
-				}, "https://api.sphere.io"),
-				Description: "CommercesTools API URL",
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("CTP_API_URL", nil),
+				Description: "The API URL of the commercetools platform. https://docs.commercetools.com/http-api",
+				Removed:     "Use the region and cloud_provider fields, to let the provider construct the correct hostname.",
 			},
-			"scopes": {
-				Type:     schema.TypeString,
-				Optional: true,
-				DefaultFunc: schema.MultiEnvDefaultFunc([]string{
-					"CTP_SCOPES",
-				}, nil),
-				Description: "CommercesTools Scopes",
+			"token_url": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("CTP_AUTH_URL", nil),
+				Description: "The authentication URL of the commercetools platform. https://docs.commercetools.com/http-api-authorization",
+				Removed:     "Use the region and cloud_provider fields, to let the provider construct the correct hostname.",
+			},
+			"region": {
+				Type:        schema.TypeString,
+				Required:    true,
```

Comment:
The new fields `region` and `cloud_provider` should be required.

Reply:
Nice! I'm wondering if we should validate the actual values, there seem to be a limited set of options right now: https://docs.commercetools.com/http-api#hosts

Reply:
The `region` and `cloud_provider` have been removed again. See https://github.com/labd/terraform-provider-commercetools/pull/107#issuecomment-590607129

Reply:
We could look into those fields again as a follow up maybe.